src/PIL/GifImagePlugin.py

```diff
@@ -824,9 +824,17 @@ def _get_optimize(im, info):
                 if count:
                     used_palette_colors.append(i)
 
+            num_palette_colors = (
+                len(im.palette.palette) // 4
+                if im.palette.mode == "RGBA"
+                else len(im.palette.palette) // 3
+            )
+            # Round up to power of 2 but at least 4
+            num_palette_colors = max(4, 1 << (num_palette_colors - 1).bit_length())
```

Comment:
Why at least 4?

Reply:
Perhaps I didn't think it through carefully enough. The idea is that this condition would trigger only when the number of used colors is low enough to reduce the GIF color table. For example, if the palette now has 30 colors, the GIF would have 32 without optimization, so if the used colors are 16 or fewer, then it is worth remapping down. So I was thinking that the right side of the `len(used_palette_colors) <= num_palette_colors // 2` condition should never be less than 2 because that's the smallest size GIF color table.

Now I see that this could mean attempting to optimize if the actual palette is 1 or 2 colors and and only one is used (an odd case but not impossible), which is probably silly. Maybe it should instead have a test to prevent optimization if there are fewer than 3 colors in the palette? Reducing a 4-color palette to 2 is the minimum reasonable case for optimization. Maybe replace

```python
            # Round up to power of 2 but at least 4
            num_palette_colors = max(4, 1 << (num_palette_colors - 1).bit_length())
```
with something like

```python
            # Round up to power of 2 (0 -> 2 but that's OK)
            num_palette_colors = 1 << (num_palette_colors - 1).bit_length()
            # Suppress if less than 4
            num_palette_colors = 0 if num_palette_colors < 4 else num_palette_colors
```

Thoughts?

Reply:
I've added another commit to https://github.com/raygard/Pillow/pull/3 with my suggestion. See what you think

Reply:
Looks good! I merged into my own repo and the changes apparently flow automatically into this PR. I'm still learning how github works...
Thank you for helping me get this working right and clean. I would not have been able to fix the tests myself.
Next up: I have mods about ready (I hope) to finally allow the "LZW minimum code size" (part 22 of the GIF spec) to be less than 8 in files written by Pillow.